**Project Name:** Supply Chain Dataset Uncleaned from DataCo Global Company


**Dataset Information:** This project analyzes a hybrid dataset from DataCo Global, capturing the end to end supply chain ecosystem. The data spans critical operational phases including provisioning, production, sales, and commercial distribution, providing a 360-degree view of the company's logistics and profitability.

**Data Architecture:**

The dataset is composed of three distinct files:

**DataCoSupplyChainDataset.csv (Structured):** The primary transactional registry containing detailed order, shipping, and customer data.

**tokenized_access_logs.csv (Unstructured):** Tokenized web logs capturing user access patterns and system interactions.

**DescriptionDataCoSupplyChain.csv (Metadata):** A data dictionary defining the variables used across the datasets.

**Aim of the Project:** To process this information, we will utilize key Python libraries including Pandas and NumPy to perform rigorous data cleaning and manipulation.

Data Source : https://www.kaggle.com/datasets/shashwatwork/dataco-smart-supply-chain-for-big-data-analysis/data

In [1]:
#Importing all the modules

import pandas as pd
import numpy as numpy
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# Loading the 3 datasets
DataCoSupplyChainDataset = pd.read_csv("/content/drive/MyDrive/Me/My project/Power BI/Datasets/Supply chain Dataset/DataCoSupplyChainDataset.csv", encoding='ISO-8859-1')
DescriptionDataCoSupplyChain = pd.read_csv("/content/drive/MyDrive/Me/My project/Power BI/Datasets/Supply chain Dataset/DescriptionDataCoSupplyChain.csv")
Tokenized_access_logs = pd.read_csv("/content/drive/MyDrive/Me/My project/Power BI/Datasets/Supply chain Dataset/tokenized_access_logs.csv")

We have 3 datasets as we know that its uncleaned and messy, may contain Null values. So we will be cleaning them individually.

In [3]:
# Setting the Column view to max.
pd.set_option('display.max_columns', None)

#### Data Cleaning (DataCoSupplyChainDataset)

In [4]:
# Viewing the top 5 rows of DataCoSupplyChainDataset
DataCoSupplyChainDataset.head()

Type  Days for shipping (real)  Days for shipment (scheduled)  \
0     DEBIT                         3                              4   
1  TRANSFER                         5                              4   
2      CASH                         4                              4   
3     DEBIT                         3                              4   
4   PAYMENT                         2                              4   

   Benefit per order  Sales per customer   Delivery Status  \
0          91.250000          314.640015  Advance shipping   
1        -249.089996          311.359985     Late delivery   
2        -247.779999          309.720001  Shipping on time   
3          22.860001          304.809998  Advance shipping   
4         134.210007          298.250000  Advance shipping   

   Late_delivery_risk  Category Id   Category Name Customer City  \
0                   0           73  Sporting Goods        Caguas   
1                   1           73  Sporting Goods        Caguas   
2                   0           73  Sporting Goods      San Jose   
3                   0           73  Sporting Goods   Los Angeles   
4                   0           73  Sporting Goods        Caguas   

  Customer Country Customer Email Customer Fname  Customer Id Customer Lname  \
0      Puerto Rico      XXXXXXXXX          Cally        20755       Holloway   
1      Puerto Rico      XXXXXXXXX          Irene        19492           Luna   
2          EE. UU.      XXXXXXXXX        Gillian        19491      Maldonado   
3          EE. UU.      XXXXXXXXX           Tana        19490           Tate   
4      Puerto Rico      XXXXXXXXX           Orli        19489      Hendricks   

  Customer Password Customer Segment Customer State           Customer Street  \
0         XXXXXXXXX         Consumer             PR  5365 Noble Nectar Island   
1         XXXXXXXXX         Consumer             PR          2679 Rustic Loop   
2         XXXXXXXXX         Consumer             CA      8510 Round Bear Gate   
3         XXXXXXXXX      Home Office             CA           3200 Amber Bend   
4         XXXXXXXXX        Corporate             PR  8671 Iron Anchor Corners   

   Customer Zipcode  Department Id Department Name   Latitude   Longitude  \
0             725.0              2         Fitness  18.251453  -66.037056   
1             725.0              2         Fitness  18.279451  -66.037064   
2           95125.0              2         Fitness  37.292233 -121.881279   
3           90027.0              2         Fitness  34.125946 -118.291016   
4             725.0              2         Fitness  18.253769  -66.037048   

         Market  Order City Order Country  Order Customer Id  \
0  Pacific Asia      Bekasi     Indonesia              20755   
1  Pacific Asia     Bikaner         India              19492   
2  Pacific Asia     Bikaner         India              19491   
3  Pacific Asia  Townsville     Australia              19490   
4  Pacific Asia  Townsville     Australia              19489   

  order date (DateOrders)  Order Id  Order Item Cardprod Id  \
0         1/31/2018 22:56     77202                    1360   
1         1/13/2018 12:27     75939                    1360   
2         1/13/2018 12:06     75938                    1360   
3         1/13/2018 11:45     75937                    1360   
4         1/13/2018 11:24     75936                    1360   

   Order Item Discount  Order Item Discount Rate  Order Item Id  \
0            13.110000                      0.04         180517   
1            16.389999                      0.05         179254   
2            18.030001                      0.06         179253   
3            22.940001                      0.07         179252   
4            29.500000                      0.09         179251   

   Order Item Product Price  Order Item Profit Ratio  Order Item Quantity  \
0                    327.75                     0.29                    1   
1                    327.75             

We can notice that there are few useless columns in the dataset and we will be dropping these columns:

1. **Customer Email** - *Filled with one static value (XXXXXXXXX).*

2. **Customer Password** - *Filled with one static value (XXXXXXXXX).*
3. **Customer Fname** - *It's not useful for the Analysis, Because Logistics/Supply-Chain Analysis add zero Analytical Value.*
4. **Customer Lname** - *It's not useful for the Analysis, Because Logistics/Supply-Chain Analysis add zero Analytical Value.*
5. **Customer Street** - *It's not useful for the Analysis, Firstly it's personal information of the Customer, so we can't use any personal information of the customer for the analysis and We have enough columns for better geographic features, such as: (Customer City, Customer State, Customer Country, Latitude, Longitude.*
6. **Customer Zipcode** - *It's not useful for the Analysis, Firstly it's personal information of the Customer, so we can't use any personal information of the customer for the analysis and We have enough columns for better geographic features, such as: (Customer City, Customer State, Customer Country, Latitude, Longitude.*
7. **Order Zipcode** - *86% of the values are missing from the column, so it's impossible to fill meaningful values as there will not be any reliable pattern. Secondly Zipcode detials is not useful for the Analysis, because we already have stronger location fields such as: (Order City, Order State, Order Country, Latitude & Longitude) as these column provide most of the required information for the geographical information.*
8. **Product Description** - *As the Entire Column is filled with Nan Values.*
9. **Product Status** - *Filled wiht one static value 0. So it's not useful for the Analysis.*

In [5]:
DataCoSupplyChainDataset.shape

(180519, 53)

We can see that their is 180519 rows and 53 columns in DataCoSupplyChainDataset.

Dropping the the Above Columns:

In [6]:
cols_to_drop = [
    'Customer Email',
    'Customer Password',
    'Customer Fname',
    'Customer Lname',
    'Customer Street',
    'Customer Zipcode',
    'Order Zipcode',
    'Product Description',
    'Product Status'
]

DataCoSupplyChainDataset.drop(columns=cols_to_drop, inplace=True)

In [7]:
DataCoSupplyChainDataset.head()

Type  Days for shipping (real)  Days for shipment (scheduled)  \
0     DEBIT                         3                              4   
1  TRANSFER                         5                              4   
2      CASH                         4                              4   
3     DEBIT                         3                              4   
4   PAYMENT                         2                              4   

   Benefit per order  Sales per customer   Delivery Status  \
0          91.250000          314.640015  Advance shipping   
1        -249.089996          311.359985     Late delivery   
2        -247.779999          309.720001  Shipping on time   
3          22.860001          304.809998  Advance shipping   
4         134.210007          298.250000  Advance shipping   

   Late_delivery_risk  Category Id   Category Name Customer City  \
0                   0           73  Sporting Goods        Caguas   
1                   1           73  Sporting Goods        Caguas   
2                   0           73  Sporting Goods      San Jose   
3                   0           73  Sporting Goods   Los Angeles   
4                   0           73  Sporting Goods        Caguas   

  Customer Country  Customer Id Customer Segment Customer State  \
0      Puerto Rico        20755         Consumer             PR   
1      Puerto Rico        19492         Consumer             PR   
2          EE. UU.        19491         Consumer             CA   
3          EE. UU.        19490      Home Office             CA   
4      Puerto Rico        19489        Corporate             PR   

   Department Id Department Name   Latitude   Longitude        Market  \
0              2         Fitness  18.251453  -66.037056  Pacific Asia   
1              2         Fitness  18.279451  -66.037064  Pacific Asia   
2              2         Fitness  37.292233 -121.881279  Pacific Asia   
3              2         Fitness  34.125946 -118.291016  Pacific Asia   
4              2         Fitness  18.253769  -66.037048  Pacific Asia   

   Order City Order Country  Order Customer Id order date (DateOrders)  \
0      Bekasi     Indonesia              20755         1/31/2018 22:56   
1     Bikaner         India              19492         1/13/2018 12:27   
2     Bikaner         India              19491         1/13/2018 12:06   
3  Townsville     Australia              19490         1/13/2018 11:45   
4  Townsville     Australia              19489         1/13/2018 11:24   

   Order Id  Order Item Cardprod Id  Order Item Discount  \
0     77202                    1360            13.110000   
1     75939                    1360            16.389999   
2     75938                    1360            18.030001   
3     75937                    1360            22.940001   
4     75936                    1360            29.500000   

   Order Item Discount Rate  Order Item Id  Order Item Product Price  \
0                      0.04         180517                    327.75   
1                      0.05         179254                    327.75   
2                      0.06         179253                    327.75   
3                      0.07         179252                    327.75   
4                      0.09         179251                    327.75   

   Order Item Profit Ratio  Order Item Quantity   Sales  Order Item Total  \
0                     0.29                    1  327.75        314.640015   
1                    -0.80                    1  327.75        311.359985   
2                    -0.80                    1  327.75        309.720001   
3                     0.08                    1  327.75        304.809998   
4                     0.45                    1  327.75        298.250000   

   Order Profit Per Order    Order Region      Order State     Order Status  \
0               91.250000  Southeast Asia  Java Occidental         COMPLETE   
1             -249.089996      South Asia         Rajastán          PENDING   
2          

In [8]:
# Checking the Number of Columns and Rows.
DataCoSupplyChainDataset.shape

(180519, 44)

We can see that their is 180519 rows and 44 columns in DataCoSupplyChainDataset.

In [9]:
# Checking for the info of the dataset
DataCoSupplyChainDataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180519 entries, 0 to 180518
Data columns (total 44 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   Type                           180519 non-null  object 
 1   Days for shipping (real)       180519 non-null  int64  
 2   Days for shipment (scheduled)  180519 non-null  int64  
 3   Benefit per order              180519 non-null  float64
 4   Sales per customer             180519 non-null  float64
 5   Delivery Status                180519 non-null  object 
 6   Late_delivery_risk             180519 non-null  int64  
 7   Category Id                    180519 non-null  int64  
 8   Category Name                  180519 non-null  object 
 9   Customer City                  180519 non-null  object 
 10  Customer Country               180519 non-null  object 
 11  Customer Id                    180519 non-null  int64  
 12  Customer Segment              

In [10]:
# Checking for Null Values
DataCoSupplyChainDataset.isnull().sum()

Type                             0
Days for shipping (real)         0
Days for shipment (scheduled)    0
Benefit per order                0
Sales per customer               0
Delivery Status                  0
Late_delivery_risk               0
Category Id                      0
Category Name                    0
Customer City                    0
Customer Country                 0
Customer Id                      0
Customer Segment                 0
Customer State                   0
Department Id                    0
Department Name                  0
Latitude                         0
Longitude                        0
Market                           0
Order City                       0
Order Country                    0
Order Customer Id                0
order date (DateOrders)          0
Order Id                         0
Order Item Cardprod Id           0
Order Item Discount              0
Order Item Discount Rate         0
Order Item Id                    0
Order Item Product Price         0
Order Item Profit Ratio          0
Order Item Quantity              0
Sales                            0
Order Item Total                 0
Order Profit Per Order           0
Order Region                     0
Order State                      0
Order Status                     0
Product Card Id                  0
Product Category Id              0
Product Image                    0
Product Name                     0
Product Price                    0
shipping date (DateOrders)       0
Shipping Mode                    0
dtype: int64

In DataCoSupplyChainDataset does'nt have Null Values.

In [11]:
DataCoSupplyChainDataset.duplicated().sum()

np.int64(0)

In DataCoSupplyChainDataset there is'nt any Duplicated Values.

So far we have removed some useless columns and checked for the null values and duplicate values, Now we will check for data quality of the dataset.

In [12]:
# Checking for the spread of the data for numerical columns.
DataCoSupplyChainDataset.describe()

Days for shipping (real)  Days for shipment (scheduled)  \
count             180519.000000                  180519.000000   
mean                   3.497654                       2.931847   
std                    1.623722                       1.374449   
min                    0.000000                       0.000000   
25%                    2.000000                       2.000000   
50%                    3.000000                       4.000000   
75%                    5.000000                       4.000000   
max                    6.000000                       4.000000   

       Benefit per order  Sales per customer  Late_delivery_risk  \
count      180519.000000       180519.000000       180519.000000   
mean           21.974989          183.107609            0.548291   
std           104.433526          120.043670            0.497664   
min         -4274.979980            7.490000            0.000000   
25%             7.000000          104.379997            0.000000   
50%            31.520000          163.990005            1.000000   
75%            64.800003          247.399994            1.000000   
max           911.799988         1939.989990            1.000000   

         Category Id    Customer Id  Department Id       Latitude  \
count  180519.000000  180519.000000  180519.000000  180519.000000   
mean       31.851451    6691.379495       5.443460      29.719955   
std        15.640064    4162.918106       1.629246       9.813646   
min         2.000000       1.000000       2.000000     -33.937553   
25%        18.000000    3258.500000       4.000000      18.265432   
50%        29.000000    6457.000000       5.000000      33.144863   
75%        45.000000    9779.000000       7.000000      39.279617   
max        76.000000   20757.000000      12.000000      48.781933   

           Longitude  Order Customer Id       Order Id  \
count  180519.000000      180519.000000  180519.000000   
mean      -84.915675        6691.379495   36221.894903   
std        21.433241        4162.918106   21045.379569   
min      -158.025986           1.000000       1.000000   
25%       -98.446312        3258.500000   18057.000000   
50%       -76.847908        6457.000000   36140.000000   
75%       -66.370583        9779.000000   54144.000000   
max       115.263077       20757.000000   77204.000000   

       Order Item Cardprod Id  Order Item Discount  Order Item Discount Rate  \
count           180519.000000        180519.000000             180519.000000   
mean               692.509764            20.664741                  0.101668   
std                336.446807            21.800901                  0.070415   
min                 19.000000             0.000000                  0.000000   
25%                403.000000             5.400000                  0.040000   
50%                627.000000            14.000000                  0.100000   
75%               1004.000000            29.990000                  0.160000   
max               1363.000000           500.000000                  0.250000   

       Order Item Id  Order Item Product Price  Order Item Profit Ratio  \
count  180519.000000             180519.000000            180519.000000   
mean    90260.000000                141.232550                 0.120647   
std     52111.490959                139.732492                 0.466796   
min         1.000000                  9.990000                -2.750000   
25%     45130.500000                 50.000000                 0.080000   
50%     90260.000000                 59.990002                 0.270000   
75%    135389.500000                199.990005                 0.360000   
max    180519.000000               1999.989990                 0.500000   

       Order Item Quantity          Sales  Order Item Total  \
count        180519.000000  180519.000000     180519.000000   
mean              2.127638     203.772096        183.107609   
std               1.453451     132.273077        120.043670   
min  

We have problem in 3 columns from the numerical columns:

1. **Benefit per order** - *has extreme high negative values of min(-4274.979980).*

2. **Order Profit Per Order** - *has extreme high negative values of min(-4274.979980).*

But we can observe that in our dataset, there is not any column that consists the cost of the product and we don't have a clear picture, how the benefit per order and order profit per order is computed, so we are going to keep the values as it is.

As we are done with the Numerical values, now we check the values in Categorical Values.

In [ ]:
# Checking for the data quality in Categorical data types

DataCoSupplyChainDataset_cols = list(DataCoSupplyChainDataset.select_dtypes(include='object').columns)

def return_unique_values(n):
  for i in n:
    print(f"{i} : {list(DataCoSupplyChainDataset[i].unique())}")
    print('-'*80)
    print('-'*80)

return_unique_values(DataCoSupplyChainDataset_cols)

We can notice that all the columns have meaningful values, Except the Order State column, Some of the state names are spelled wrong.
For Example: Rajastán, \x8eilina. I think its a encoding problem, so we will be using unidecode module to fix these problem.

In [14]:
# Installing the unidecode Module

!pip install unidecode

In [15]:
import unidecode

DataCoSupplyChainDataset['Order State Clean'] = DataCoSupplyChainDataset['Order State'].apply(unidecode.unidecode)

In [16]:
list(DataCoSupplyChainDataset['Order State Clean'].unique())

['Java Occidental',
 'Rajastan',
 'Queensland',
 'Guangdong',
 'Tokio',
 'Celebes Septentrional',
 'Maharashtra',
 'Seul',
 'Madhya Pradesh',
 'Victoria',
 'Australia Occidental',
 'Guangxi',
 'Delhi',
 'Gansu',
 'Singapur',
 'Karnataka',
 'Nueva Gales del Sur',
 'Sumatra Septentrional',
 'Territorio de la Capital Australiana',
 'Zhejiang',
 'Uttar Pradesh',
 'Bursa',
 'Australia del Sur',
 'Estambul',
 'Ulan Bator',
 'Washington',
 'California',
 'Kansas',
 'Cross River',
 'Kinshasa',
 'Dakar',
 'Gran Casablanca',
 'Renania del Norte-Westfalia',
 'Isla de Francia',
 'Paises del Loira',
 'Groningen',
 'Inglaterra',
 'Alsacia-Champana-Ardenas-Lorena',
 'Provenza-Alpes-Costa Azul',
 'Guatemala',
 'San Salvador',
 'Chiriqui',
 'Santo Domingo',
 'Zulia',
 'Risaralda',
 'Francisco Morazan',
 'Bahia',
 'Distrito Federal',
 'Jalisco',
 'Puebla',
 'Sao Paulo',
 'Montevideo',
 'Buenos Aires',
 'Parana',
 'Camaguey',
 'Las Tunas',
 'Nayarit',
 'Panama',
 'Mexico',
 'Alagoas',
 'Minas Gerais',
 '

As we can see that, it has fixed, but there are certain state names, still its not fixed. we will be fixing it manually using the mapping.

In [17]:
# Defining the Error State Name in a Dictionaries.
fix_map = {
    'Kahramanmara?': 'Kahramanmaras',
    'Th? Do Ha N?i': 'Hanoi',
    'T?nh C?n Th?': 'Can Tho',
    'Klaip?da': 'Klaipeda',
    'Gy?r': 'Gyor',
    'Bra?ov': 'Brasov',
    'iauliai': 'Siauliai',
    'ilina': 'Zilina',
    'Kahramanmara?': 'Kahramanmaras',

    # Remove redundant qualifiers
    'Lima (ciudad)': 'Lima',
    'Sofia-Ciudad': 'Sofia',
    'Basilea-Ciudad': 'Basel-Stadt',
    'Departamento Atlantico': 'Atlantico',
    'Cidade De Maputo': 'Maputo',
    'Grad Zagreb': 'Zagreb',
    'Gran Casablanca': 'Casablanca',
    'Buyumbura Mairie': 'Bujumbura Mairie',

    # Specific translations
    'Franzosisch-Guayana': 'French Guiana'
}

# Applying the Dictionaries,
DataCoSupplyChainDataset['Order State Clean'] = DataCoSupplyChainDataset['Order State Clean'].replace(fix_map)

In [22]:
# Ensure all names are String Datatype
DataCoSupplyChainDataset['Order State Clean'] = DataCoSupplyChainDataset['Order State Clean'].astype(str)

# Removing the unwanted Space
DataCoSupplyChainDataset['Order State Clean'] = DataCoSupplyChainDataset['Order State Clean'].str.strip()

In [23]:
list(DataCoSupplyChainDataset['Order State Clean'].unique())

['Java Occidental',
 'Rajastan',
 'Queensland',
 'Guangdong',
 'Tokio',
 'Celebes Septentrional',
 'Maharashtra',
 'Seul',
 'Madhya Pradesh',
 'Victoria',
 'Australia Occidental',
 'Guangxi',
 'Delhi',
 'Gansu',
 'Singapur',
 'Karnataka',
 'Nueva Gales del Sur',
 'Sumatra Septentrional',
 'Territorio de la Capital Australiana',
 'Zhejiang',
 'Uttar Pradesh',
 'Bursa',
 'Australia del Sur',
 'Estambul',
 'Ulan Bator',
 'Washington',
 'California',
 'Kansas',
 'Cross River',
 'Kinshasa',
 'Dakar',
 'Casablanca',
 'Renania del Norte-Westfalia',
 'Isla de Francia',
 'Paises del Loira',
 'Groningen',
 'Inglaterra',
 'Alsacia-Champana-Ardenas-Lorena',
 'Provenza-Alpes-Costa Azul',
 'Guatemala',
 'San Salvador',
 'Chiriqui',
 'Santo Domingo',
 'Zulia',
 'Risaralda',
 'Francisco Morazan',
 'Bahia',
 'Distrito Federal',
 'Jalisco',
 'Puebla',
 'Sao Paulo',
 'Montevideo',
 'Buenos Aires',
 'Parana',
 'Camaguey',
 'Las Tunas',
 'Nayarit',
 'Panama',
 'Mexico',
 'Alagoas',
 'Minas Gerais',
 'Santi

Now we have fixed all the State Names.

In [26]:
DataCoSupplyChainDataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180519 entries, 0 to 180518
Data columns (total 45 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   Type                           180519 non-null  object        
 1   Days for shipping (real)       180519 non-null  int64         
 2   Days for shipment (scheduled)  180519 non-null  int64         
 3   Benefit per order              180519 non-null  float64       
 4   Sales per customer             180519 non-null  float64       
 5   Delivery Status                180519 non-null  object        
 6   Late_delivery_risk             180519 non-null  int64         
 7   Category Id                    180519 non-null  int64         
 8   Category Name                  180519 non-null  object        
 9   Customer City                  180519 non-null  object        
 10  Customer Country               180519 non-null  object        
 11  

We can notice that order date (DateOrders) and shipping date (DateOrders) is in float type. we should change it to datetime datatype.

In [25]:
DataCoSupplyChainDataset['order date (DateOrders)'] = pd.to_datetime(DataCoSupplyChainDataset['order date (DateOrders)'])
DataCoSupplyChainDataset['shipping date (DateOrders)'] = pd.to_datetime(DataCoSupplyChainDataset['shipping date (DateOrders)'])

In [131]:
DataCoSupplyChainDataset['order date (DateOrders)'] = DataCoSupplyChainDataset['order date (DateOrders)'].dt.strftime('%d-%m-%Y')
DataCoSupplyChainDataset['shipping date (DateOrders)'] = DataCoSupplyChainDataset['shipping date (DateOrders)'].dt.strftime('%d-%m-%Y')

In [132]:
DataCoSupplyChainDataset[['order date (DateOrders)', 'shipping date (DateOrders)']].sample(10)

order date (DateOrders) shipping date (DateOrders)
10830               04-05-2017                 06-05-2017
136663              11-07-2017                 13-07-2017
57814               27-09-2017                 01-10-2017
137551              22-05-2017                 24-05-2017
54348               19-06-2017                 21-06-2017
59699               20-07-2017                 26-07-2017
73531               24-11-2016                 28-11-2016
70671               26-01-2018                 28-01-2018
164160              23-10-2015                 27-10-2015
139827              26-05-2016                 31-05-2016

In [39]:
DataCoSupplyChainDataset.head()

Type  Days for shipping (real)  Days for shipment (scheduled)  \
0     DEBIT                         3                              4   
1  TRANSFER                         5                              4   
2      CASH                         4                              4   
3     DEBIT                         3                              4   
4   PAYMENT                         2                              4   

   Benefit per order  Sales per customer   Delivery Status  \
0          91.250000          314.640015  Advance shipping   
1        -249.089996          311.359985     Late delivery   
2        -247.779999          309.720001  Shipping on time   
3          22.860001          304.809998  Advance shipping   
4         134.210007          298.250000  Advance shipping   

   Late_delivery_risk  Category Id   Category Name Customer City  \
0                   0           73  Sporting Goods        Caguas   
1                   1           73  Sporting Goods        Caguas   
2                   0           73  Sporting Goods      San Jose   
3                   0           73  Sporting Goods   Los Angeles   
4                   0           73  Sporting Goods        Caguas   

  Customer Country  Customer Id Customer Segment Customer State  \
0      Puerto Rico        20755         Consumer             PR   
1      Puerto Rico        19492         Consumer             PR   
2          EE. UU.        19491         Consumer             CA   
3          EE. UU.        19490      Home Office             CA   
4      Puerto Rico        19489        Corporate             PR   

   Department Id Department Name   Latitude   Longitude        Market  \
0              2         Fitness  18.251453  -66.037056  Pacific Asia   
1              2         Fitness  18.279451  -66.037064  Pacific Asia   
2              2         Fitness  37.292233 -121.881279  Pacific Asia   
3              2         Fitness  34.125946 -118.291016  Pacific Asia   
4              2         Fitness  18.253769  -66.037048  Pacific Asia   

   Order City Order Country  Order Customer Id order date (DateOrders)  \
0      Bekasi     Indonesia              20755     2018-01-31 22:56:00   
1     Bikaner         India              19492     2018-01-13 12:27:00   
2     Bikaner         India              19491     2018-01-13 12:06:00   
3  Townsville     Australia              19490     2018-01-13 11:45:00   
4  Townsville     Australia              19489     2018-01-13 11:24:00   

   Order Id  Order Item Cardprod Id  Order Item Discount  \
0     77202                    1360            13.110000   
1     75939                    1360            16.389999   
2     75938                    1360            18.030001   
3     75937                    1360            22.940001   
4     75936                    1360            29.500000   

   Order Item Discount Rate  Order Item Id  Order Item Product Price  \
0                      0.04         180517                    327.75   
1                      0.05         179254                    327.75   
2                      0.06         179253                    327.75   
3                      0.07         179252                    327.75   
4                      0.09         179251                    327.75   

   Order Item Profit Ratio  Order Item Quantity   Sales  Order Item Total  \
0                     0.29                    1  327.75        314.640015   
1                    -0.80                    1  327.75        311.359985   
2                    -0.80                    1  327.75        309.720001   
3                     0.08                    1  327.75        304.809998   
4                     0.45                    1  327.75        298.250000   

   Order Profit Per Order    Order Region      Order State     Order Status  \
0               91.250000  Southeast Asia  Java Occidental         COMPLETE   
1             -249.089996      South Asia         Rajastán          PENDING   
2          

In [43]:
# Rearranging the Column order

New_order_columns = ['Type', 'Days for shipping (real)', 'Days for shipment (scheduled)',
       'Benefit per order', 'Sales per customer', 'Delivery Status',
       'Late_delivery_risk', 'Category Id', 'Category Name', 'Customer City',
       'Customer Country', 'Customer Id', 'Customer Segment', 'Customer State',
       'Department Id', 'Department Name', 'Latitude', 'Longitude', 'Market',
       'Order City', 'Order Country', 'Order Customer Id',
       'order date (DateOrders)', 'Order Id', 'Order Item Cardprod Id',
       'Order Item Discount', 'Order Item Discount Rate', 'Order Item Id',
       'Order Item Product Price', 'Order Item Profit Ratio',
       'Order Item Quantity', 'Sales', 'Order Item Total',
       'Order Profit Per Order', 'Order Region', 'Order State', 'Order State Clean', 'Order Status',
       'Product Card Id', 'Product Category Id', 'Product Image',
       'Product Name', 'Product Price', 'shipping date (DateOrders)',
       'Shipping Mode']

DataCoSupplyChainDataset = DataCoSupplyChainDataset[New_order_columns]

In [44]:
DataCoSupplyChainDataset.head()

Type  Days for shipping (real)  Days for shipment (scheduled)  \
0     DEBIT                         3                              4   
1  TRANSFER                         5                              4   
2      CASH                         4                              4   
3     DEBIT                         3                              4   
4   PAYMENT                         2                              4   

   Benefit per order  Sales per customer   Delivery Status  \
0          91.250000          314.640015  Advance shipping   
1        -249.089996          311.359985     Late delivery   
2        -247.779999          309.720001  Shipping on time   
3          22.860001          304.809998  Advance shipping   
4         134.210007          298.250000  Advance shipping   

   Late_delivery_risk  Category Id   Category Name Customer City  \
0                   0           73  Sporting Goods        Caguas   
1                   1           73  Sporting Goods        Caguas   
2                   0           73  Sporting Goods      San Jose   
3                   0           73  Sporting Goods   Los Angeles   
4                   0           73  Sporting Goods        Caguas   

  Customer Country  Customer Id Customer Segment Customer State  \
0      Puerto Rico        20755         Consumer             PR   
1      Puerto Rico        19492         Consumer             PR   
2          EE. UU.        19491         Consumer             CA   
3          EE. UU.        19490      Home Office             CA   
4      Puerto Rico        19489        Corporate             PR   

   Department Id Department Name   Latitude   Longitude        Market  \
0              2         Fitness  18.251453  -66.037056  Pacific Asia   
1              2         Fitness  18.279451  -66.037064  Pacific Asia   
2              2         Fitness  37.292233 -121.881279  Pacific Asia   
3              2         Fitness  34.125946 -118.291016  Pacific Asia   
4              2         Fitness  18.253769  -66.037048  Pacific Asia   

   Order City Order Country  Order Customer Id order date (DateOrders)  \
0      Bekasi     Indonesia              20755     2018-01-31 22:56:00   
1     Bikaner         India              19492     2018-01-13 12:27:00   
2     Bikaner         India              19491     2018-01-13 12:06:00   
3  Townsville     Australia              19490     2018-01-13 11:45:00   
4  Townsville     Australia              19489     2018-01-13 11:24:00   

   Order Id  Order Item Cardprod Id  Order Item Discount  \
0     77202                    1360            13.110000   
1     75939                    1360            16.389999   
2     75938                    1360            18.030001   
3     75937                    1360            22.940001   
4     75936                    1360            29.500000   

   Order Item Discount Rate  Order Item Id  Order Item Product Price  \
0                      0.04         180517                    327.75   
1                      0.05         179254                    327.75   
2                      0.06         179253                    327.75   
3                      0.07         179252                    327.75   
4                      0.09         179251                    327.75   

   Order Item Profit Ratio  Order Item Quantity   Sales  Order Item Total  \
0                     0.29                    1  327.75        314.640015   
1                    -0.80                    1  327.75        311.359985   
2                    -0.80                    1  327.75        309.720001   
3                     0.08                    1  327.75        304.809998   
4                     0.45                    1  327.75        298.250000   

   Order Profit Per Order    Order Region      Order State Order State Clean  \
0               91.250000  Southeast Asia  Java Occidental   Java Occidental   
1             -249.089996      South Asia         Rajastán          Rajastan   
2       

Now we have cleaned the DataCoSupplyChainDataset. Let's move to the Tokenized_access_logs Dataset.

In [46]:
Tokenized_access_logs.head()

Product             Category  \
0     adidas Brazuca 2017 Official Match Ball  baseball & softball   
1       The North Face Women's Recon Backpack   hunting & shooting   
2      adidas Kids' RG III Mid Football Cleat       featured shops   
3  Under Armour Men's Compression EV SL Slide          electronics   
4                 Pelican Sunstream 100 Kayak         water sports   

            Date Month  Hour Department              ip  \
0  9/1/2017 6:00   Sep     6   fitness     37.97.182.65   
1  9/1/2017 6:00   Sep     6  fan shop     206.56.112.1   
2  9/1/2017 6:00   Sep     6   apparel    215.143.180.0   
3  9/1/2017 6:00   Sep     6  footwear     206.56.112.1   
4  9/1/2017 6:01   Sep     6  fan shop   136.108.56.242   

                                                 url  
0  /department/fitness/category/baseball%20&%20so...  
1  /department/fan%20shop/category/hunting%20&%20...  
2  /department/apparel/category/featured%20shops/...  
3  /department/footwear/category/electronics/prod...  
4  /department/fan%20shop/category/water%20sports...

In [45]:
Tokenized_access_logs.shape

(469977, 8)

We can see that their is 469977 rows and 8 columns in Tokenized_access_logs.

In [97]:
# Checking info of the dataset
Tokenized_access_logs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 469977 entries, 0 to 469976
Data columns (total 8 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   Product     469977 non-null  object
 1   Category    469977 non-null  object
 2   Date        469977 non-null  object
 3   Month       469977 non-null  object
 4   Hour        469977 non-null  int64 
 5   Department  469977 non-null  object
 6   ip          469977 non-null  object
 7   url         469977 non-null  object
dtypes: int64(1), object(7)
memory usage: 28.7+ MB


We can notice that Date column is in Object Data Type and the datetime format is mixed, so we will be standardise it to "%d-%m-%Y" format, Rest of the columns are fine.

In [101]:
Tokenized_access_logs['Date'] = pd.to_datetime(Tokenized_access_logs['Date'], dayfirst=True, errors='coerce')
Tokenized_access_logs['Date'] = pd.to_datetime(Tokenized_access_logs['Date'], dayfirst=True, errors='coerce').dt.strftime('%d-%m-%Y')

In [102]:
# Checking the Date Column
Tokenized_access_logs['Date'].sample(20)

124578    11-09-2017
294980    14-09-2017
165646    19-12-2017
115947    28-10-2017
48084     16-10-2017
33999     04-12-2017
181165    21-12-2017
411376    23-01-2018
450419    02-10-2017
97232     25-10-2017
345734    17-09-2017
358636    13-01-2018
42150     06-09-2017
211593    14-09-2017
159157    13-09-2017
6692      02-09-2017
146992    03-11-2017
272028    23-11-2017
13907     02-12-2017
276551    14-09-2017
Name: Date, dtype: object

We have converted the date column values into '%d-%m-%Y' Format Succesfully.

In [99]:
# Checking info of the dataset
Tokenized_access_logs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 469977 entries, 0 to 469976
Data columns (total 8 columns):
 #   Column      Non-Null Count   Dtype         
---  ------      --------------   -----         
 0   Product     469977 non-null  object        
 1   Category    469977 non-null  object        
 2   Date        469977 non-null  datetime64[ns]
 3   Month       469977 non-null  object        
 4   Hour        469977 non-null  int64         
 5   Department  469977 non-null  object        
 6   ip          469977 non-null  object        
 7   url         469977 non-null  object        
dtypes: datetime64[ns](1), int64(1), object(6)
memory usage: 28.7+ MB


In [21]:
# Checking for Null Values
Tokenized_access_logs.isnull().sum()

Product       0
Category      0
Date          0
Month         0
Hour          0
Department    0
ip            0
url           0
dtype: int64

There is Null Values in the dataset, that's a very good thing.

In [128]:
# Checking the data in random rows
Tokenized_access_logs.sample(10)

Product         Category  \
278312      Field & Stream Sportsman 16 Gun Fire Safe          fishing   
137303      Team Golf St. Louis Cardinals Putter Grip      accessories   
11607         Nike Men's Free TR 5.0 TB Training Shoe  as seen on  tv!   
408585  adidas Youth Germany Black/Red Away Match Soc   girls' apparel   
171513           Nike Men's Dri-FIT Victory Golf Polo  women's apparel   
264520  Diamondback Boys' Insight 24 Performance Hybr       basketball   
392180        Nike Men's CJ Elite 2 TD Football Cleat   men's footwear   
285373                         Bag Boy M330 Push Cart      golf gloves   
125338        Nike Men's Free TR 5.0 TB Training Shoe  as seen on  tv!   
265308      Field & Stream Sportsman 16 Gun Fire Safe          fishing   

              Date Month  Hour Department              ip  \
278312  24-11-2017   Nov     6  fan shop   14.161.219.217   
137303  01-11-2017   Nov    14  outdoors   143.182.50.145   
11607   11-10-2017   Oct    20  footwear    212.26.127.49   
408585  26-09-2017   Sep    15      golf    192.105.2.226   
171513  14-09-2017   Sep    11      golf   135.247.44.107   
264520  14-09-2017   Sep    21   fitness      82.77.81.68   
392180  24-09-2017   Sep     8   apparel    189.18.197.88   
285373  03-01-2018   Jan    15  outdoors   116.164.189.38   
125338  30-10-2017   Oct    11  footwear   159.21.136.111   
265308  14-09-2017   Sep    21  fan shop   194.172.187.58   

                                                      url  
278312  /department/fan%20shop/category/fishing/produc...  
137303  /department/outdoors/category/accessories/prod...  
11607   /department/footwear/category/as%20seen%20on%2...  
408585  /department/golf/category/girls'%20apparel/pro...  
171513  /department/golf/category/women's%20apparel/pr...  
264520  /department/fitness/category/basketball/produc...  
392180  /department/apparel/category/men's%20footwear/...  
285373  /department/outdoors/category/golf%20gloves/pr...  
125338  /department/footwear/category/as%20seen%20on%2...  
265308  /department/fan%20shop/category/fishing/produc...

The Dataset now has been cleaned, let's export the dataset which we have cleaned and load it in Power BI for Visualization.

In [133]:
# Exporting the DataCoSupplyChainDataset to Excel Format

DataCoSupplyChainDataset.to_excel("DataCoSupplyChainDataset_cleaned.xlsx", index=False)

In [130]:
# Exporting the Tokenized_access_logs to Excel Format

Tokenized_access_logs.to_excel("Tokenized_access_logs.xlsx", index=False)